tests/consensus_tests/utils.py

```diff
@@ -26,12 +49,17 @@ def every_test():
 
 
 def get_port() -> int:
-    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
-        s.bind(('', 0))
-        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-        return s.getsockname()[1]
-
-
+    while True:
+        with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
+            # get random port assigned by the OS
+            s.bind(('', 0))
```

Comment:
if we try to bind to an already bound port it will crash with an exception

Reply:
but i think that this assumption about getting already allocated port from `s.bind(('', 0))` should be double-checked, i wouldn't expect it to behave this way actually

Reply:
I read a bit about this and the OS will pick a random, unallocated port, so in theory, this could happen if the OS runs out of ports, but then we have a different problem to deal with.

Reply:
Some sources:
https://stackoverflow.com/a/1365284
https://www.dnorth.net/2012/03/17/the-port-0-trick/

Reply:
@kwkr check the latest CI run, there is an error regarding a collision for hardcoded ports I believe 
https://github.com/qdrant/qdrant/actions/runs/6572224277/job/17852947397?pr=2831

Reply:
The problem was:
```
busy_ports[self.http_port] = None
busy_ports[self.grpc_port] = None
busy_ports[self.p2p_port] = None
```

which doesn't actually remove a key from a dictionary in python. I already fixed this and running the tests locally to confirm.

Reply:
@kwkr I want to make sure we are fixing the right problem here :)

Assuming that:
- `get_port` is not called concurrently
- `s.bind(('', 0))` returns only unallocated ports.
- `with closing(...)` ensures that the port is released when returning the port to the caller of `get_port`

It seems that the only explanation why we get port conflict is that we spawn server processes very quickly without giving them necessarily the time to bind on their assigned ports.

Meaning serial calls to `get_port` could returned the same port.

Am I getting this right?

Reply:
With the implementation that I did, the port that is about to be allocated is added to the `busy_ports` right away, so calling `get_port` in a sequence shouldn't return a port, that was already returned. Only after the process is killed, does the port get removed from there and can be returned again. In the current implementation, calling `get_port` in a sequence can indeed return the same port, because there is a pretty significant delay between picking the port and the process starting using the port. The processes are spawned one after another so in the test scenarios that are there, the `get_port` call can happen without a port being allocated in the meantime even 12 times (in `test_points_recommendation.py` , 5 peers in total, waiting for one leader and then spawning 4 peers one after another, each needing 3 ports). It's not that likely to happen, but given how many tests there are, the probability gets bigger.

If I got this correctly, when getting 12 ports in a row, this happens on average every 230 situations and for 6 ports, every 910 situations. So given the overall traffic in your tests it can sometimes happen as the failed runs show.

To be sure, I ran this script to count it:
```python
import socket
from contextlib import closing

def get_port() -> int:
    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
        # get random port assigned by the OS
        s.bind(('', 0))
        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        allocated_port = s.getsockname()[1]
        return allocated_port

def test_ports():
    for k in range(0, 10000):
        ports_to_allocate = 6
        s = set()
        for _ in range(0,ports_to_allocate):
            s.add(get_port())
        if len(s) != ports_to_allocate:
            print(len(s))
            return k
```